pallets/services-payment/src/lib.rs

```diff
@@ -342,16 +290,54 @@ impl<T: Config> AuthorNotingHook<T::AccountId> for Pallet<T> {
         _block_number: BlockNumber,
         para_id: ParaId,
     ) -> Weight {
-        let total_weight = T::DbWeight::get().reads_writes(1, 1);
+        if Pallet::<T>::burn_credit_for_para(&para_id).is_err() {
+            let (amount_to_charge, _weight) = T::ProvideBlockProductionCost::block_cost(&para_id);
+            match T::Currency::withdraw(
+                &Self::parachain_tank(para_id),
+                amount_to_charge,
+                WithdrawReasons::FEE,
+                ExistenceRequirement::AllowDeath,
+            ) {
+                Err(e) => log::warn!(
+                    "Failed to withdraw credits for container chain {}: {:?}",
+                    u32::from(para_id),
+                    e
+                ),
+                Ok(imbalance) => {
+                    T::OnChargeForBlock::on_unbalanced(imbalance);
+                }
+            }
+        }
 
-        if let Err(e) = Pallet::<T>::burn_credit_for_para(&para_id) {
-            log::warn!(
-                "Failed to burn credits for container chain {}: {:?}",
-                u32::from(para_id),
-                e
-            );
+        T::WeightInfo::on_container_author_noted()
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Derive a derivative account ID from the paraId.
+    pub fn parachain_tank(para_id: ParaId) -> T::AccountId {
+        let entropy = (b"modlpy/serpayment", para_id).using_encoded(blake2_256);
+        Decode::decode(&mut TrailingZeroInput::new(entropy.as_ref()))
+            .expect("infinite length input; no invalid inputs for type; qed")
+    }
+
+    /// Hook to perform things on deregister
+    pub fn para_deregistered(para_id: ParaId) {
+        // Drain the para-id account from tokens
+        let parachain_tank_balance = T::Currency::total_balance(&Self::parachain_tank(para_id));
+        if !parachain_tank_balance.is_zero() {
+            if let Ok(imbalance) = T::Currency::withdraw(
+                &Self::parachain_tank(para_id),
+                parachain_tank_balance,
+                WithdrawReasons::FEE,
+                ExistenceRequirement::AllowDeath,
+            ) {
+                // Burn for now, we might be able to pass something to do with this
+                drop(imbalance);
```

Comment:
Is there a specific reason why this imbalance is not handled using `OnChargeForBlock` type as the one it is inside `on_container_author_noted` hook function? I mean like `T::OnChargeForBlock::on_unbalanced(imbalance)`.

Reply:
I guess due to `OnChargeForBlock` is just for fees handling? But I wonder if we could use the same mechanism to deal with the unspent tokens from the tank.

Reply:
The plan is to sent them to a user-configured account, but that will be implemented in a future PR because it may require a refactor of `pallet_registrar`.

Reply:
yes exactly, I did not want to make this PR bigger than it already is (we also have an issue for the follow up where we refund the unspent tank tokens)